components/widgets/menu/menu_items/submenu_item.jsx

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import * as Utils from 'utils/utils.jsx';
+import {ModalIdentifiers} from 'utils/constants';
+import SubMenuModal from '../menu_modals/submenu_modal/submenu_modal.jsx';
+import {openModal} from 'actions/views/modals';
+import store from 'stores/redux_store';
+
+// Requires an object conforming to a submenu structure passed to registerPostDropdownMenuAction
+// of the form:
+// {
+//     "id": "A",
+//     "text": "A text",
+//     "subMenu": [
+//         {
+//             "id": "B",
+//             "text": "B text"
+//             "action": () => {},
+//             "filter": () => {},
+//         }
+//     ],
+//     "action": () => {},
+//     "filter": () => {},
+// }
+// Submenus can contain Submenus as well
+export default class SubMenuItem extends React.PureComponent {
+    static propTypes= {
+        id: PropTypes.string,
+        postId: PropTypes.string,
+        text: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
+        subMenu: PropTypes.arrayOf(PropTypes.object),
+        icon: PropTypes.node,
+        action: PropTypes.func,
+        filter: PropTypes.func,
+        xOffset: PropTypes.number,
+        ariaLabel: PropTypes.string,
+        root: PropTypes.bool,
+    };
+    static defaultProps = {
+        show: true,
+    };
+
+    constructor(props) {
+        super(props);
+        this.node = React.createRef();
+
+        this.state = {
+            show: false,
+        };
+    }
+
+    show = () => {
+        this.setState({show: true});
+    }
+
+    hide = () => {
+        this.setState({show: false});
+    }
+
+    handleShowMobileSubMenuItem = (elements) => {
+        const submenuModalData = {
+            ModalId: ModalIdentifiers.MOBILE_SUBMENU,
+            dialogType: SubMenuModal,
+            dialogProps: {
+                elements,
+            },
+        };
+
+        store.dispatch(openModal(submenuModalData));
```

Comment:
I removed store + modal dependencies from `submenuitem`, and put the same logic in `actions/global_actions.jsx` in `showMobileSubMenuModal`.  Thoughts?

Reply:
Hmm not sure if that is optimal either. I was hoping that either we can inject the action when the `SubMenu` is created or create another component that is redux aware that manages the state. It might be difficult to talk about it here, maybe we can hop on a call if possible and discuss a solution?

Reply:
Sounds good.  Will ping you on community.

Reply:
@ali-farooq0 , were you looking to see something like `actions.openModal(modalData)` passed in as a prop?  Your recommendation of rebuilding SubMenuIteam to be like `MenuItemToggleModalRedux` is good, however I honestly feel out of scope for the PR.  Can that be included in further work as it would be an improvement?